### PR TITLE
Proof of concept: parameterized node traversal methods.

### PR DESCRIPTION
**Pull request by frewsxcv**
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/14451)

### CODE REVIEW

**Comment by frewsxcv** [comment]
I've had this sitting around in a branch for a while, thought I might as well open a PR. There's still a little more work to do here, but I didn't want to spend any more time on it if the idea is not well received.

**Comment by jdm** [comment]
I'm in favour!

**Comment by Ms2ger** [comment]
Can we use default type parameters yet?

**Comment by frewsxcv** [comment]
> Can we use default type parameters yet?

I thought about making it `....<T=Node>(..)`, but then I ran into https://github.com/rust-lang/rust/issues/36887

**Comment by frewsxcv** [comment]
More info: https://internals.rust-lang.org/t/interaction-of-user-defined-and-integral-fallbacks-with-inference/2496

File: components/script/lib.rs [hunk 1]
```diff
@@ -2,6 +2,8 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
+#![allow(unrooted_must_root)]
```
Comment (jdm) [inline components/script/lib.rs]: Please remove this. We should only scope it for the code that requires it.

**Comment by Wafflespeanut** [comment]
Ping?

**Comment by frewsxcv** [comment]
I don't have much time right now to finish this up. If someone else wants to, go for it :)

**Comment by jdm** [comment]
Tracking this work in https://github.com/servo/servo/issues/15270.

**Pull request closed without merging**